In [2]:
# Vanguard A/B Testing Project
## 01 — Data Discovery & Cleaning

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [4]:
demo = pd.read_csv("../data/raw/df_final_demo.txt", sep=",")
experiment_clients = pd.read_csv("../data/raw/df_final_experiment_clients.txt", sep=",")
web_data_pt1 = pd.read_csv("../data/raw/df_final_web_data_pt_1.txt", sep=",")
web_data_pt2 = pd.read_csv("../data/raw/df_final_web_data_pt_2.txt", sep=",")

In [194]:
web_data = pd.concat(
    [web_data_pt1, web_data_pt2],
    ignore_index=True
)

print("Web data shape:", web_data.shape)
print("Unique clients:", web_data["client_id"].nunique())
print("\nColumns:")
print(web_data.columns.tolist())

Web data shape: (755405, 5)
Unique clients: 120157

Columns:
['client_id', 'visitor_id', 'visit_id', 'process_step', 'date_time']


In [293]:
web_with_variation = web_data.merge(
    experiment_clients[["client_id", "Variation"]],
    on="client_id",
    how="inner"
)

print("Shape:", web_with_variation.shape)
print("\nVariation:")
print(web_with_variation["Variation"].value_counts(dropna=False))

Shape: (449831, 6)

Variation:
Variation
Test       177847
Control    143462
NaN        128522
Name: count, dtype: int64


In [294]:
print(demo.shape)
print(experiment_clients.shape)
print(web_data_pt1.shape)
print(web_data_pt2.shape)

(70609, 9)
(70609, 2)
(343141, 5)
(412264, 5)


In [295]:
print(demo.columns.tolist())
print(experiment_clients.columns.tolist())
print(web_data_pt1.columns.tolist())
print(web_data_pt2.columns.tolist())

['client_id', 'clnt_tenure_yr', 'clnt_tenure_mnth', 'clnt_age', 'gendr', 'num_accts', 'bal', 'calls_6_mnth', 'logons_6_mnth']
['client_id', 'Variation']
['client_id', 'visitor_id', 'visit_id', 'process_step', 'date_time']
['client_id', 'visitor_id', 'visit_id', 'process_step', 'date_time']


In [296]:
missing_variation = experiment_clients[
    experiment_clients["Variation"].isna()
].copy()

missing_client_ids = set(missing_variation["client_id"])

print("Clients with missing Variation: - 01_data_discovery_and_cleaning.ipynb:7", len(missing_variation))

Clients with missing Variation: - 01_data_discovery_and_cleaning.ipynb:7 20109


In [297]:
# Combine cleaned web data
web_data_pt1_clean = web_data_pt1.drop_duplicates().copy()
web_data_pt2_clean = web_data_pt2.drop_duplicates().copy()

web_data_clean = pd.concat(
    [web_data_pt1_clean, web_data_pt2_clean],
    ignore_index=True
)

# Identify clients without an A/B assignment
missing_web_data = web_data_clean[
    web_data_clean["client_id"].isin(missing_client_ids)
].copy()

In [298]:
missing_web_data["process_step"]

243851    confirm
243852     step_3
243853     step_2
243854     step_1
243855      start
           ...   
744613     step_3
744614     step_2
744615     step_1
744616      start
744617      start
Name: process_step, Length: 126662, dtype: str

In [299]:
print("Total web events: - 01_data_discovery_and_cleaning.ipynb:1", len(missing_web_data))

print("\nProcess steps: - 01_data_discovery_and_cleaning.ipynb:3")
print(missing_web_data.iloc[:, 3].value_counts())

print("\nUnique clients: - 01_data_discovery_and_cleaning.ipynb:6")
print(missing_web_data.iloc[:, 0].nunique())

Total web events: - 01_data_discovery_and_cleaning.ipynb:1 126662

Process steps: - 01_data_discovery_and_cleaning.ipynb:3
process_step
start      40263
step_1     27297
step_2     22594
step_3     19323
confirm    17185
Name: count, dtype: int64

Unique clients: - 01_data_discovery_and_cleaning.ipynb:6
20109


In [300]:
web_data_pt1_clean = web_data_pt1.drop_duplicates().copy()
web_data_pt2_clean = web_data_pt2.drop_duplicates().copy()

web_data_clean = pd.concat(
    [web_data_pt1_clean, web_data_pt2_clean],
    ignore_index=True
)

In [301]:
missing_web_data.iloc[:, 3].value_counts()

process_step
start      40263
step_1     27297
step_2     22594
step_3     19323
confirm    17185
Name: count, dtype: int64

In [302]:
demo.head()

,client_id,clnt_tenure_yr,clnt_tenure_mnth,clnt_age,gendr,num_accts,bal,calls_6_mnth,logons_6_mnth
0,836976,6.0,73.0,60.5,U,2.0,45105.30,6.0,9.0
1,2304905,7.0,94.0,58.0,U,2.0,110860.30,6.0,9.0
2,1439522,5.0,64.0,32.0,U,2.0,52467.79,6.0,9.0
3,1562045,16.0,198.0,49.0,M,2.0,67454.65,3.0,6.0
4,5126305,12.0,145.0,33.0,F,2.0,103671.75,0.0,3.0


In [303]:
experiment_clients.head()

,client_id,Variation
0,9988021,Test
1,8320017,Test
2,4033851,Control
3,1982004,Test
4,9294070,Control


In [304]:
web_data_pt1.head()

,client_id,visitor_id,visit_id,process_step,date_time
0,9988021,580560515_7732621733,781255054_21935453173_531117,step_3,2017-04-17 15:27:07
1,9988021,580560515_7732621733,781255054_21935453173_531117,step_2,2017-04-17 15:26:51
2,9988021,580560515_7732621733,781255054_21935453173_531117,step_3,2017-04-17 15:19:22
3,9988021,580560515_7732621733,781255054_21935453173_531117,step_2,2017-04-17 15:19:13
4,9988021,580560515_7732621733,781255054_21935453173_531117,step_3,2017-04-17 15:18:04


In [305]:
print(demo.shape)
print(experiment_clients.shape)
print(web_data_pt1.shape)
print(web_data_pt2.shape)

(70609, 9)
(70609, 2)
(343141, 5)
(412264, 5)


In [306]:
print("Demo: - 01_data_discovery_and_cleaning.ipynb:1", demo.shape)
print("Experiment clients: - 01_data_discovery_and_cleaning.ipynb:2", experiment_clients.shape)
print("Web data part 1: - 01_data_discovery_and_cleaning.ipynb:3", web_data_pt1.shape)
print("Web data part 2: - 01_data_discovery_and_cleaning.ipynb:4", web_data_pt2.shape)

Demo: - 01_data_discovery_and_cleaning.ipynb:1 (70609, 9)
Experiment clients: - 01_data_discovery_and_cleaning.ipynb:2 (70609, 2)
Web data part 1: - 01_data_discovery_and_cleaning.ipynb:3 (343141, 5)
Web data part 2: - 01_data_discovery_and_cleaning.ipynb:4 (412264, 5)


In [307]:
print(demo.columns.tolist())
print(experiment_clients.columns.tolist())
print(web_data_pt1.columns.tolist())
print(web_data_pt2.columns.tolist())

['client_id', 'clnt_tenure_yr', 'clnt_tenure_mnth', 'clnt_age', 'gendr', 'num_accts', 'bal', 'calls_6_mnth', 'logons_6_mnth']
['client_id', 'Variation']
['client_id', 'visitor_id', 'visit_id', 'process_step', 'date_time']
['client_id', 'visitor_id', 'visit_id', 'process_step', 'date_time']


In [308]:
print("=== DEMO DATA TYPES === - 01_data_discovery_and_cleaning.ipynb:1")
print(demo.dtypes)

print("\n=== EXPERIMENT CLIENTS DATA TYPES === - 01_data_discovery_and_cleaning.ipynb:4")
print(experiment_clients.dtypes)

print("\n=== WEB DATA PART 1 DATA TYPES === - 01_data_discovery_and_cleaning.ipynb:7")
print(web_data_pt1.dtypes)

print("\n=== WEB DATA PART 2 DATA TYPES === - 01_data_discovery_and_cleaning.ipynb:10")
print(web_data_pt2.dtypes)

=== DEMO DATA TYPES === - 01_data_discovery_and_cleaning.ipynb:1
client_id             int64
clnt_tenure_yr      float64
clnt_tenure_mnth    float64
clnt_age            float64
gendr                   str
num_accts           float64
bal                 float64
calls_6_mnth        float64
logons_6_mnth       float64
dtype: object

=== EXPERIMENT CLIENTS DATA TYPES === - 01_data_discovery_and_cleaning.ipynb:4
client_id    int64
Variation      str
dtype: object

=== WEB DATA PART 1 DATA TYPES === - 01_data_discovery_and_cleaning.ipynb:7
client_id       int64
visitor_id        str
visit_id          str
process_step      str
date_time         str
dtype: object

=== WEB DATA PART 2 DATA TYPES === - 01_data_discovery_and_cleaning.ipynb:10
client_id       int64
visitor_id        str
visit_id          str
process_step      str
date_time         str
dtype: object


In [309]:
# Check missing values

print("=== DEMO MISSING VALUES === - 01_data_discovery_and_cleaning.ipynb:3")
print(demo.isnull().sum())

print("\n=== EXPERIMENT CLIENTS MISSING VALUES === - 01_data_discovery_and_cleaning.ipynb:6")
print(experiment_clients.isnull().sum())

print("\n=== WEB DATA PART 1 MISSING VALUES === - 01_data_discovery_and_cleaning.ipynb:9")
print(web_data_pt1.isnull().sum())

print("\n=== WEB DATA PART 2 MISSING VALUES === - 01_data_discovery_and_cleaning.ipynb:12")
print(web_data_pt2.isnull().sum())

=== DEMO MISSING VALUES === - 01_data_discovery_and_cleaning.ipynb:3
client_id            0
clnt_tenure_yr      14
clnt_tenure_mnth    14
clnt_age            15
gendr               14
num_accts           14
bal                 14
calls_6_mnth        14
logons_6_mnth       14
dtype: int64

=== EXPERIMENT CLIENTS MISSING VALUES === - 01_data_discovery_and_cleaning.ipynb:6
client_id        0
Variation    20109
dtype: int64

=== WEB DATA PART 1 MISSING VALUES === - 01_data_discovery_and_cleaning.ipynb:9
client_id       0
visitor_id      0
visit_id        0
process_step    0
date_time       0
dtype: int64

=== WEB DATA PART 2 MISSING VALUES === - 01_data_discovery_and_cleaning.ipynb:12
client_id       0
visitor_id      0
visit_id        0
process_step    0
date_time       0
dtype: int64


In [310]:
# Check rows with missing demographic information

demo[demo.isnull().any(axis=1)]

,client_id,clnt_tenure_yr,clnt_tenure_mnth,clnt_age,gendr,num_accts,bal,calls_6_mnth,logons_6_mnth
4164,7402828,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
8316,355337,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
8677,8412164,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
9583,4666211,8.0,106.0,NaN,F,2.0,42550.55,4.0,7.0
13444,2222915,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
18066,4876926,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
25961,5277910,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
28432,7616759,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
35323,8191345,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
43518,1227228,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [311]:
# Check duplicate rows

print("Demo duplicates: - 01_data_discovery_and_cleaning.ipynb:3", demo.duplicated().sum())
print("Experiment duplicates: - 01_data_discovery_and_cleaning.ipynb:4", experiment_clients.duplicated().sum())
print("Web data part 1 duplicates: - 01_data_discovery_and_cleaning.ipynb:5", web_data_pt1.duplicated().sum())
print("Web data part 2 duplicates: - 01_data_discovery_and_cleaning.ipynb:6", web_data_pt2.duplicated().sum())

Demo duplicates: - 01_data_discovery_and_cleaning.ipynb:3 0
Experiment duplicates: - 01_data_discovery_and_cleaning.ipynb:4 0
Web data part 1 duplicates: - 01_data_discovery_and_cleaning.ipynb:5 2095
Web data part 2 duplicates: - 01_data_discovery_and_cleaning.ipynb:6 8669


In [312]:
experiment_clients.iloc[:, 1].isna().sum()

np.int64(20109)

In [313]:
# Inspect exact duplicates in web data part 1

duplicates_pt1 = web_data_pt1[
    web_data_pt1.duplicated(keep=False)
]

duplicates_pt1.head(20)

,client_id,visitor_id,visit_id,process_step,date_time
364,9358626,148475529_63668223960,223297395_36250329195_832161,start,2017-04-28 12:27:28
365,9358626,148475529_63668223960,223297395_36250329195_832161,start,2017-04-28 12:27:28
367,5376894,622509328_44408201172,688984457_43441834354_912755,start,2017-04-28 14:20:01
368,5376894,622509328_44408201172,688984457_43441834354_912755,start,2017-04-28 14:20:01
382,438567,923318467_50708963627,330543236_29863358529_771432,start,2017-04-28 02:52:29
383,438567,923318467_50708963627,330543236_29863358529_771432,start,2017-04-28 02:52:29
784,1423857,716830992_81930238382,568819166_80596727019_75889,confirm,2017-04-10 22:44:58
785,1423857,716830992_81930238382,568819166_80596727019_75889,confirm,2017-04-10 22:44:58
1815,372523,730241475_25771921122,25331354_74479847014_854987,start,2017-04-27 10:51:26
1816,372523,730241475_25771921122,25331354_74479847014_854987,start,2017-04-27 10:51:26


In [314]:
# Inspect exact duplicates in web data part 1

duplicates_pt1 = web_data_pt1[
    web_data_pt1.duplicated(keep=False)
]

duplicates_pt1.head(20)

,client_id,visitor_id,visit_id,process_step,date_time
364,9358626,148475529_63668223960,223297395_36250329195_832161,start,2017-04-28 12:27:28
365,9358626,148475529_63668223960,223297395_36250329195_832161,start,2017-04-28 12:27:28
367,5376894,622509328_44408201172,688984457_43441834354_912755,start,2017-04-28 14:20:01
368,5376894,622509328_44408201172,688984457_43441834354_912755,start,2017-04-28 14:20:01
382,438567,923318467_50708963627,330543236_29863358529_771432,start,2017-04-28 02:52:29
383,438567,923318467_50708963627,330543236_29863358529_771432,start,2017-04-28 02:52:29
784,1423857,716830992_81930238382,568819166_80596727019_75889,confirm,2017-04-10 22:44:58
785,1423857,716830992_81930238382,568819166_80596727019_75889,confirm,2017-04-10 22:44:58
1815,372523,730241475_25771921122,25331354_74479847014_854987,start,2017-04-27 10:51:26
1816,372523,730241475_25771921122,25331354_74479847014_854987,start,2017-04-27 10:51:26


In [315]:
# Inspect exact duplicates in web data part 2

duplicates_pt2 = web_data_pt2[
    web_data_pt2.duplicated(keep=False)
]

duplicates_pt2.head(20)

,client_id,visitor_id,visit_id,process_step,date_time
124,3751049,937925261_74821699083,628817270_45049432060_475841,start,2017-05-15 23:09:44
125,3751049,937925261_74821699083,628817270_45049432060_475841,start,2017-05-15 23:09:44
142,9912156,746623513_52101087617,780044921_57678038107_984942,start,2017-05-03 14:09:52
143,9912156,746623513_52101087617,780044921_57678038107_984942,start,2017-05-03 14:09:52
325,5786602,893139880_61416949586,747153834_69434320996_858907,start,2017-05-10 01:49:09
326,5786602,893139880_61416949586,747153834_69434320996_858907,start,2017-05-10 01:49:09
327,5786602,893139880_61416949586,747153834_69434320996_858907,start,2017-05-10 01:48:45
328,5786602,893139880_61416949586,747153834_69434320996_858907,start,2017-05-10 01:48:45
374,5786602,893139880_61416949586,119968876_5423296886_390511,start,2017-06-08 02:14:09
375,5786602,893139880_61416949586,119968876_5423296886_390511,start,2017-06-08 02:14:09


In [316]:
# Inspect exact duplicates in web data part 2

duplicates_pt2 = web_data_pt2[
    web_data_pt2.duplicated(keep=False)
]

duplicates_pt2.head(20)

,client_id,visitor_id,visit_id,process_step,date_time
124,3751049,937925261_74821699083,628817270_45049432060_475841,start,2017-05-15 23:09:44
125,3751049,937925261_74821699083,628817270_45049432060_475841,start,2017-05-15 23:09:44
142,9912156,746623513_52101087617,780044921_57678038107_984942,start,2017-05-03 14:09:52
143,9912156,746623513_52101087617,780044921_57678038107_984942,start,2017-05-03 14:09:52
325,5786602,893139880_61416949586,747153834_69434320996_858907,start,2017-05-10 01:49:09
326,5786602,893139880_61416949586,747153834_69434320996_858907,start,2017-05-10 01:49:09
327,5786602,893139880_61416949586,747153834_69434320996_858907,start,2017-05-10 01:48:45
328,5786602,893139880_61416949586,747153834_69434320996_858907,start,2017-05-10 01:48:45
374,5786602,893139880_61416949586,119968876_5423296886_390511,start,2017-06-08 02:14:09
375,5786602,893139880_61416949586,119968876_5423296886_390511,start,2017-06-08 02:14:09


In [317]:
print("Part 1 duplicated events: - 01_data_discovery_and_cleaning.ipynb:1", duplicates_pt1.value_counts().head(10))
print("Part 2 duplicated events: - 01_data_discovery_and_cleaning.ipynb:2", duplicates_pt2.value_counts().head(10))

Part 1 duplicated events: - 01_data_discovery_and_cleaning.ipynb:1 client_id  visitor_id             visit_id                      process_step  date_time          
7699482    816681289_34976118382  598603763_85483459636_600247  start         2017-04-03 09:04:09    6
3176103    973929305_26287939905  698510236_6551485027_519901   confirm       2017-04-16 16:38:40    4
7706345    171294152_14292105319  20317641_22107870639_335169   step_3        2017-04-06 07:37:50    4
5951709    480620449_45705876248  635448583_60628693020_361150  step_1        2017-04-19 11:27:53    4
5427647    613809412_97249620846  339824895_73123393347_159684  step_2        2017-03-29 13:17:40    4
8955845    851442620_43576576883  830957511_35452452087_927337  confirm       2017-04-05 11:51:23    4
7699482    816681289_34976118382  598603763_85483459636_600247  start         2017-04-03 09:04:12    4
2153749    142434497_22880557373  112944941_42863832160_218959  confirm       2017-04-18 14:33:34    4
8390041    

In [318]:
# Remove exact duplicate rows

web_data_pt1_clean = web_data_pt1.drop_duplicates().copy()
web_data_pt2_clean = web_data_pt2.drop_duplicates().copy()

print("Web data part 1: - 01_data_discovery_and_cleaning.ipynb:6")
print("Before: - 01_data_discovery_and_cleaning.ipynb:7", web_data_pt1.shape)
print("After: - 01_data_discovery_and_cleaning.ipynb:8", web_data_pt1_clean.shape)

print("\nWeb data part 2: - 01_data_discovery_and_cleaning.ipynb:10")
print("Before: - 01_data_discovery_and_cleaning.ipynb:11", web_data_pt2.shape)
print("After: - 01_data_discovery_and_cleaning.ipynb:12", web_data_pt2_clean.shape)

Web data part 1: - 01_data_discovery_and_cleaning.ipynb:6
Before: - 01_data_discovery_and_cleaning.ipynb:7 (343141, 5)
After: - 01_data_discovery_and_cleaning.ipynb:8 (341046, 5)

Web data part 2: - 01_data_discovery_and_cleaning.ipynb:10
Before: - 01_data_discovery_and_cleaning.ipynb:11 (412264, 5)
After: - 01_data_discovery_and_cleaning.ipynb:12 (403595, 5)


In [319]:
print("Part 1 duplicates remaining: - 01_data_discovery_and_cleaning.ipynb:1", web_data_pt1_clean.duplicated().sum())
print("Part 2 duplicates remaining: - 01_data_discovery_and_cleaning.ipynb:2", web_data_pt2_clean.duplicated().sum())

Part 1 duplicates remaining: - 01_data_discovery_and_cleaning.ipynb:1 0
Part 2 duplicates remaining: - 01_data_discovery_and_cleaning.ipynb:2 0


In [320]:
# Investigate clients with missing A/B test assignment

missing_variation = experiment_clients[
    experiment_clients["Variation"].isna()
].copy()

print("Clients with missing Variation: - 01_data_discovery_and_cleaning.ipynb:7", len(missing_variation))

print("\nFirst 10 clients with missing Variation: - 01_data_discovery_and_cleaning.ipynb:9")
display(missing_variation.head(10))

print("\nDo these clients exist in the demographic dataset? - 01_data_discovery_and_cleaning.ipynb:12")
print(
    missing_variation["client_id"].isin(demo["client_id"]).value_counts()
)

Clients with missing Variation: - 01_data_discovery_and_cleaning.ipynb:7 20109

First 10 clients with missing Variation: - 01_data_discovery_and_cleaning.ipynb:9


,client_id,Variation
50500,5459747,NaN
50501,8031000,NaN
50502,1847030,NaN
50503,9713157,NaN
50504,7775828,NaN
50505,7644910,NaN
50506,5945107,NaN
50507,2212145,NaN
50508,9755145,NaN
50509,5129355,NaN



Do these clients exist in the demographic dataset? - 01_data_discovery_and_cleaning.ipynb:12
client_id
True    20109
Name: count, dtype: int64


In [321]:
missing_variation["client_id"].isin(demo["client_id"]).value_counts()

client_id
True    20109
Name: count, dtype: int64

In [322]:
missing_variation["client_id"].isin(demo["client_id"]).value_counts()

client_id
True    20109
Name: count, dtype: int64

In [323]:
# Check whether clients without an A/B assignment appear in the web data

missing_client_ids = set(missing_variation["client_id"])

web_client_ids = set(
    pd.concat([
        web_data_pt1_clean["client_id"],
        web_data_pt2_clean["client_id"]
    ]).unique()
)

missing_clients_in_web = missing_client_ids.intersection(web_client_ids)

print("Clients with missing Variation: - 01_data_discovery_and_cleaning.ipynb:14", len(missing_client_ids))
print("MissingVariation clients found in web data: - 01_data_discovery_and_cleaning.ipynb:15", len(missing_clients_in_web))
print(
    "Missing-Variation clients NOT found in web data:",
    len(missing_client_ids - web_client_ids)
)

Clients with missing Variation: - 01_data_discovery_and_cleaning.ipynb:14 20109
MissingVariation clients found in web data: - 01_data_discovery_and_cleaning.ipynb:15 20109
Missing-Variation clients NOT found in web data: 0


In [324]:
import pandas as pd

In [325]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [326]:
print(demo.shape)
print(experiment_clients.shape)
print(web_data_pt1.shape)
print(web_data_pt2.shape)

(70609, 9)
(70609, 2)
(343141, 5)
(412264, 5)


In [327]:
print(demo.shape)
print(experiment_clients.shape)
print(web_data_pt1.shape)
print(web_data_pt2.shape)

(70609, 9)
(70609, 2)
(343141, 5)
(412264, 5)


In [328]:
print(demo.columns.tolist())
print(experiment_clients.columns.tolist())
print(web_data_pt1.columns.tolist())
print(web_data_pt2.columns.tolist())

['client_id', 'clnt_tenure_yr', 'clnt_tenure_mnth', 'clnt_age', 'gendr', 'num_accts', 'bal', 'calls_6_mnth', 'logons_6_mnth']
['client_id', 'Variation']
['client_id', 'visitor_id', 'visit_id', 'process_step', 'date_time']
['client_id', 'visitor_id', 'visit_id', 'process_step', 'date_time']


In [329]:
"process_step"

'process_step'

In [330]:
missing_web_data["process_step"].value_counts()

process_step
start      40263
step_1     27297
step_2     22594
step_3     19323
confirm    17185
Name: count, dtype: int64

In [331]:
print("Part 1 clean shape: - 01_data_discovery_and_cleaning.ipynb:1", web_data_pt1_clean.shape)
print("Part 2 clean shape: - 01_data_discovery_and_cleaning.ipynb:2", web_data_pt2_clean.shape)
print("Combined clean shape: - 01_data_discovery_and_cleaning.ipynb:3", web_data_clean.shape)

Part 1 clean shape: - 01_data_discovery_and_cleaning.ipynb:1 (341046, 5)
Part 2 clean shape: - 01_data_discovery_and_cleaning.ipynb:2 (403595, 5)
Combined clean shape: - 01_data_discovery_and_cleaning.ipynb:3 (744641, 5)


In [332]:
print("Missing values: - 01_data_discovery_and_cleaning.ipynb:1")
print(missing_web_data.isnull().sum())

Missing values: - 01_data_discovery_and_cleaning.ipynb:1
client_id       0
visitor_id      0
visit_id        0
process_step    0
date_time       0
dtype: int64


In [333]:
missing_web_data["date_time"] = pd.to_datetime(
    missing_web_data["date_time"]
)

In [334]:
print(missing_web_data["date_time"].dtype)
print(missing_web_data["date_time"].min())
print(missing_web_data["date_time"].max())

datetime64[us]
2017-03-15 00:03:03
2017-06-20 23:44:20


In [335]:
missing_web_data.head()

,client_id,visitor_id,visit_id,process_step,date_time
243851,7775828,995150070_44942900577,669119128_68248839401_799567,confirm,2017-03-30 01:31:25
243852,7775828,995150070_44942900577,669119128_68248839401_799567,step_3,2017-03-30 01:30:18
243853,7775828,995150070_44942900577,669119128_68248839401_799567,step_2,2017-03-30 01:29:21
243854,7775828,995150070_44942900577,669119128_68248839401_799567,step_1,2017-03-30 01:28:18
243855,7775828,995150070_44942900577,669119128_68248839401_799567,start,2017-03-30 01:28:10


In [336]:
print("Invalid dates: - 01_data_discovery_and_cleaning.ipynb:1", missing_web_data.iloc[:, 4].isna().sum())

Invalid dates: - 01_data_discovery_and_cleaning.ipynb:1 0


In [337]:
print(missing_web_data.dtypes)

client_id                int64
visitor_id                 str
visit_id                   str
process_step               str
date_time       datetime64[us]
dtype: object


In [338]:
print("Duplicate rows: - 01_data_discovery_and_cleaning.ipynb:1", missing_web_data.duplicated().sum())

Duplicate rows: - 01_data_discovery_and_cleaning.ipynb:1 0


In [339]:
missing_web_data["process_step"]

243851    confirm
243852     step_3
243853     step_2
243854     step_1
243855      start
           ...   
744613     step_3
744614     step_2
744615     step_1
744616      start
744617      start
Name: process_step, Length: 126662, dtype: str

In [340]:
print(missing_web_data["process_step"].value_counts())

process_step
start      40263
step_1     27297
step_2     22594
step_3     19323
confirm    17185
Name: count, dtype: int64


In [341]:
print("Unique clients:", missing_web_data["client_id"].nunique())

Unique clients: 20109


In [342]:
print("Duplicate rows:", missing_web_data.duplicated().sum())

Duplicate rows: 0


In [343]:
print(missing_web_data.dtypes)

client_id                int64
visitor_id                 str
visit_id                   str
process_step               str
date_time       datetime64[us]
dtype: object


In [344]:
print(missing_web_data.isnull().sum())

client_id       0
visitor_id      0
visit_id        0
process_step    0
date_time       0
dtype: int64


In [345]:
print(missing_web_data["process_step"].unique())

<StringArray>
['confirm', 'step_3', 'step_2', 'step_1', 'start']
Length: 5, dtype: str


In [346]:
print(missing_web_data["client_id"].dtype)
print(missing_web_data["client_id"].min())
print(missing_web_data["client_id"].max())

int64
169
9999839


In [347]:
print(missing_web_data["visitor_id"].dtype)
print(missing_web_data["visit_id"].dtype)
print(missing_web_data["visitor_id"].isna().sum())
print(missing_web_data["visit_id"].isna().sum())

str
str
0
0


In [348]:
print(missing_web_data["process_step"].dtype)
print(missing_web_data["process_step"].isna().sum())

str
0


In [349]:
print(missing_web_data["date_time"].dtype)
print(missing_web_data["date_time"].isna().sum())
print(missing_web_data["date_time"].min())
print(missing_web_data["date_time"].max())

datetime64[us]
0
2017-03-15 00:03:03
2017-06-20 23:44:20


In [350]:
print(
    missing_web_data
    .groupby("visit_id")["process_step"]
    .apply(list)
    .head(10)
)

visit_id
100061345_677868293_127105                                                [start]
100065579_58979991824_496649             [confirm, step_3, step_2, step_1, start]
100120278_95743019653_642729                                   [confirm, confirm]
100124986_29622874577_527522                                              [start]
100140480_55852036378_404718    [step_3, step_2, step_2, step_3, step_3, step_...
100154767_60541320490_163313    [confirm, step_3, step_2, step_1, start, step_...
100165170_70624318404_468222             [confirm, step_3, step_2, step_1, start]
100215152_57349471249_380824    [confirm, step_3, step_2, step_3, step_2, step...
100219420_22495116984_915662                 [start, start, start, step_1, start]
100250179_91945603763_367112                                       [start, start]
Name: process_step, dtype: object


In [351]:
missing_web_data_sorted = missing_web_data.sort_values(
    ["visit_id", "date_time"]
).copy()

print(
    missing_web_data_sorted
    .groupby("visit_id")["process_step"]
    .apply(list)
    .head(10)
)

visit_id
100061345_677868293_127105                                                [start]
100065579_58979991824_496649             [start, step_1, step_2, step_3, confirm]
100120278_95743019653_642729                                   [confirm, confirm]
100124986_29622874577_527522                                              [start]
100140480_55852036378_404718    [start, step_1, step_2, step_3, step_2, step_3...
100154767_60541320490_163313    [start, step_1, start, step_1, step_2, step_3,...
100165170_70624318404_468222             [start, step_1, step_2, step_3, confirm]
100215152_57349471249_380824    [start, step_1, step_2, step_3, step_2, step_3...
100219420_22495116984_915662                 [start, step_1, start, start, start]
100250179_91945603763_367112                                       [start, start]
Name: process_step, dtype: object


In [352]:
step_order = {
    "start": 0,
    "step_1": 1,
    "step_2": 2,
    "step_3": 3,
    "confirm": 4
}

def has_backward_step(steps):
    positions = [step_order[step] for step in steps]
    return any(positions[i] < positions[i - 1] for i in range(1, len(positions)))

visit_sequences = (
    missing_web_data_sorted
    .groupby("visit_id")["process_step"]
    .apply(list)
)

backward_visits = visit_sequences.apply(has_backward_step)

print("Total visits:", len(visit_sequences))
print("Visits with backward steps:", backward_visits.sum())

Total visits: 27757
Visits with backward steps: 6573


In [353]:
completed_visits = visit_sequences.apply(lambda steps: "confirm" in steps)

print("Completed visits:", completed_visits.sum())
print("Incomplete visits:", (~completed_visits).sum())

Completed visits: 15189
Incomplete visits: 12568


In [354]:
funnel = (
    missing_web_data_sorted
    .groupby("process_step")["visit_id"]
    .nunique()
    .reindex(["start", "step_1", "step_2", "step_3", "confirm"])
)

print(funnel)

process_step
start      25601
step_1     20668
step_2     17832
step_3     16222
confirm    15189
Name: visit_id, dtype: int64


In [355]:
funnel_conversion = (funnel / funnel.iloc[0] * 100).round(2)

print(funnel_conversion)

process_step
start      100.00
step_1      80.73
step_2      69.65
step_3      63.36
confirm     59.33
Name: visit_id, dtype: float64


In [356]:
print(experiment_clients["Variation"].value_counts())

Variation
Test       26968
Control    23532
Name: count, dtype: int64


In [357]:
web_with_variation = missing_web_data_sorted.merge(
    experiment_clients,
    on="client_id",
    how="left"
)

print(web_with_variation.shape)
print(web_with_variation["Variation"].value_counts(dropna=False))

(126662, 6)
Variation
NaN    126662
Name: count, dtype: int64


In [358]:
print("Web client_id dtype:", missing_web_data_sorted["client_id"].dtype)
print("Experiment client_id dtype:", experiment_clients["client_id"].dtype)

print("Web unique clients:", missing_web_data_sorted["client_id"].nunique())
print("Experiment unique clients:", experiment_clients["client_id"].nunique())

print(
    "Matching client IDs:",
    missing_web_data_sorted["client_id"]
    .isin(experiment_clients["client_id"])
    .sum()
)

Web client_id dtype: int64
Experiment client_id dtype: int64
Web unique clients: 20109
Experiment unique clients: 70609
Matching client IDs: 126662


In [359]:
web_with_variation = missing_web_data_sorted.merge(
    experiment_clients[["client_id", "Variation"]],
    on="client_id",
    how="left"
)

print(web_with_variation.shape)
print(web_with_variation["Variation"].value_counts(dropna=False))

(126662, 6)
Variation
NaN    126662
Name: count, dtype: int64


In [360]:
print(experiment_clients.head())

print(
    experiment_clients[
        experiment_clients["client_id"].isin(
            missing_web_data_sorted["client_id"].head(5)
        )
    ]
)

   client_id Variation
0    9988021      Test
1    8320017      Test
2    4033851   Control
3    1982004      Test
4    9294070   Control
       client_id Variation
62195    1748647       NaN
66508    2848373       NaN


In [361]:
test_merge = missing_web_data_sorted[
    ["client_id"]
].merge(
    experiment_clients[["client_id", "Variation"]],
    on="client_id",
    how="left",
    indicator=True
)

print(test_merge["_merge"].value_counts())
print(test_merge.head(10))

_merge
both          126662
left_only          0
right_only         0
Name: count, dtype: int64
   client_id Variation _merge
0    1748647       NaN   both
1    2848373       NaN   both
2    2848373       NaN   both
3    2848373       NaN   both
4    2848373       NaN   both
5    2848373       NaN   both
6    4456733       NaN   both
7    4456733       NaN   both
8    9156914       NaN   both
9    8964191       NaN   both


In [362]:
print("Experiment clients:", experiment_clients["client_id"].nunique())

print("Clients with Variation:")
print(experiment_clients["Variation"].notna().sum())

print("Clients without Variation:")
print(experiment_clients["Variation"].isna().sum())

print("\nVariation distribution:")
print(experiment_clients["Variation"].value_counts(dropna=False))

Experiment clients: 70609
Clients with Variation:
50500
Clients without Variation:
20109

Variation distribution:
Variation
Test       26968
Control    23532
NaN        20109
Name: count, dtype: int64


In [363]:
print(
    experiment_clients[
        experiment_clients["Variation"].isna()
    ].head(10)
)

       client_id Variation
50500    5459747       NaN
50501    8031000       NaN
50502    1847030       NaN
50503    9713157       NaN
50504    7775828       NaN
50505    7644910       NaN
50506    5945107       NaN
50507    2212145       NaN
50508    9755145       NaN
50509    5129355       NaN


In [364]:
missing_variation_ids = set(
    experiment_clients.loc[
        experiment_clients["Variation"].isna(),
        "client_id"
    ]
)

print("Missing Variation clients:", len(missing_variation_ids))

print(
    "Same as missing_variation:",
    missing_variation_ids == set(missing_variation["client_id"])
)

Missing Variation clients: 20109
Same as missing_variation: True


In [365]:
print(
    experiment_clients[
        experiment_clients["Variation"].isna()
    ]["client_id"].isin(
        missing_web_data_sorted["client_id"]
    ).sum()
)

20109


In [366]:
print("Missing Variation clients:", len(missing_variation_ids))

print(
    "Unique clients in web data:",
    missing_web_data_sorted["client_id"].nunique()
)

print(
    "All web clients have missing Variation:",
    set(missing_web_data_sorted["client_id"]).issubset(
        missing_variation_ids
    )
)

Missing Variation clients: 20109
Unique clients in web data: 20109
All web clients have missing Variation: True


In [367]:
web_missing_variation = missing_web_data_sorted.copy()

print("Web data shape:", web_missing_variation.shape)
print("Unique clients:", web_missing_variation["client_id"].nunique())
print("Missing values:")
print(web_missing_variation.isna().sum())

Web data shape: (126662, 5)
Unique clients: 20109
Missing values:
client_id       0
visitor_id      0
visit_id        0
process_step    0
date_time       0
dtype: int64


In [368]:
print(web_missing_variation.head())

        client_id             visitor_id                      visit_id  \
300576    1748647  413751600_31279009983    100061345_677868293_127105   
320955    2848373   66514700_31923440179  100065579_58979991824_496649   
320954    2848373   66514700_31923440179  100065579_58979991824_496649   
320953    2848373   66514700_31923440179  100065579_58979991824_496649   
320952    2848373   66514700_31923440179  100065579_58979991824_496649   

       process_step           date_time  
300576        start 2017-04-04 06:57:49  
320955        start 2017-04-15 14:27:50  
320954       step_1 2017-04-15 14:31:37  
320953       step_2 2017-04-15 14:32:21  
320952       step_3 2017-04-15 14:32:45  


In [369]:
web_missing_variation

,client_id,visitor_id,visit_id,process_step,date_time
300576,1748647,413751600_31279009983,100061345_677868293_127105,start,2017-04-04 06:57:49
320955,2848373,66514700_31923440179,100065579_58979991824_496649,start,2017-04-15 14:27:50
320954,2848373,66514700_31923440179,100065579_58979991824_496649,step_1,2017-04-15 14:31:37
320953,2848373,66514700_31923440179,100065579_58979991824_496649,step_2,2017-04-15 14:32:21
320952,2848373,66514700_31923440179,100065579_58979991824_496649,step_3,2017-04-15 14:32:45
...,...,...,...,...,...
329016,9219912,197363615_49244520494,999953421_18770252152_92364,step_1,2017-04-03 08:15:58
329015,9219912,197363615_49244520494,999953421_18770252152_92364,step_2,2017-04-03 08:16:06
274313,8374531,339629093_15796635997,999986292_55082159911_561557,confirm,2017-04-25 18:31:17
274312,8374531,339629093_15796635997,999986292_55082159911_561557,confirm,2017-04-25 19:02:52


In [370]:
funnel_missing_variation = (
    web_missing_variation
    .groupby("process_step")["visit_id"]
    .nunique()
    .reindex(["start", "step_1", "step_2", "step_3", "confirm"])
)

print(funnel_missing_variation)

process_step
start      25601
step_1     20668
step_2     17832
step_3     16222
confirm    15189
Name: visit_id, dtype: int64


In [371]:
funnel_missing_conversion = (
    funnel_missing_variation / funnel_missing_variation.iloc[0] * 100
).round(2)

print(funnel_missing_conversion)

process_step
start      100.00
step_1      80.73
step_2      69.65
step_3      63.36
confirm     59.33
Name: visit_id, dtype: float64


In [372]:
visits_per_client = (
    web_missing_variation
    .groupby("client_id")["visit_id"]
    .nunique()
)

print("Average visits per client:", visits_per_client.mean().round(2))
print("Median visits per client:", visits_per_client.median())
print("Maximum visits by one client:", visits_per_client.max())

print("\nVisit distribution:")
print(visits_per_client.value_counts().sort_index())

Average visits per client: 1.38
Median visits per client: 1.0
Maximum visits by one client: 21

Visit distribution:
visit_id
1     14865
2      3805
3       919
4       294
5       119
6        50
7        24
8        17
9         6
10        5
11        1
13        1
16        1
17        1
21        1
Name: count, dtype: int64


In [373]:
print(
    "Clients with 1 visit:",
    (visits_per_client == 1).sum()
)

print(
    "Clients with multiple visits:",
    (visits_per_client > 1).sum()
)

Clients with 1 visit: 14865
Clients with multiple visits: 5244


In [374]:
repeat_clients = visits_per_client[visits_per_client > 1]

print("Repeat clients:", len(repeat_clients))
print("\nMost frequent visitors:")
print(repeat_clients.sort_values(ascending=False).head(10))

Repeat clients: 5244

Most frequent visitors:
client_id
2128341    21
4925232    17
1771611    16
9663519    13
9511606    11
3676166    10
3539274    10
8223974    10
1786617    10
9497505    10
Name: visit_id, dtype: int64


In [375]:
max_visit_client = visits_per_client.idxmax()

print("Client with most visits:", max_visit_client)

print(
    web_missing_variation[
        web_missing_variation["client_id"] == max_visit_client
    ].sort_values(["visit_id", "date_time"])
)

Client with most visits: 2128341
        client_id             visitor_id                      visit_id  \
331905    2128341  690243746_97686781248  147595654_58891556188_298785   
331904    2128341  690243746_97686781248  147595654_58891556188_298785   
331954    2128341  923252049_38877790223   17495337_67678584436_866770   
331953    2128341  923252049_38877790223   17495337_67678584436_866770   
331871    2128341  585028710_71643026491  206225469_90859667128_113497   
...           ...                    ...                           ...   
331923    2128341  368926294_90402015035   98932995_77077592109_484998   
331956    2128341  368926294_90402015035   98932995_77077592109_484998   
331955    2128341  368926294_90402015035   98932995_77077592109_484998   
331952    2128341  368926294_90402015035   98932995_77077592109_484998   
331951    2128341  368926294_90402015035   98932995_77077592109_484998   

       process_step           date_time  
331905        start 2017-03-29 15:00

In [376]:
client_data = web_missing_variation[
    web_missing_variation["client_id"] == max_visit_client
]

visit_counts = (
    client_data
    .groupby("visit_id")
    .size()
    .sort_values(ascending=False)
)

print("Rows per visit:")
print(visit_counts)

Rows per visit:
visit_id
748958041_38511308408_739982    13
300576108_9655535007_130343     11
98932995_77077592109_484998     10
743032041_8924526567_601424      7
788476829_52035711559_646564     5
217878853_64864519993_211275     4
894106_67200918696_599604        4
17495337_67678584436_866770      2
147595654_58891556188_298785     2
213853768_63590088391_429330     2
206225469_90859667128_113497     2
440265138_22787625182_840683     2
388533651_44737161264_462550     2
839416575_47905599901_784019     2
690501382_19217870082_30385      2
521113182_14655971304_625433     2
495453972_30946404421_30753      2
88124520_48953096022_422523      2
821194309_33974167145_311653     2
870214894_60784041185_600579     2
508715527_23771860978_884862     1
dtype: int64


In [377]:
largest_visit = visit_counts.index[0]

print("Largest visit:", largest_visit)

print(
    client_data[
        client_data["visit_id"] == largest_visit
    ].sort_values("date_time")
)

Largest visit: 748958041_38511308408_739982
        client_id             visitor_id                      visit_id  \
331903    2128341  923252049_38877790223  748958041_38511308408_739982   
331902    2128341  923252049_38877790223  748958041_38511308408_739982   
331901    2128341  923252049_38877790223  748958041_38511308408_739982   
331900    2128341  923252049_38877790223  748958041_38511308408_739982   
331899    2128341  923252049_38877790223  748958041_38511308408_739982   
331898    2128341  923252049_38877790223  748958041_38511308408_739982   
331897    2128341  923252049_38877790223  748958041_38511308408_739982   
331896    2128341  923252049_38877790223  748958041_38511308408_739982   
331895    2128341  923252049_38877790223  748958041_38511308408_739982   
331892    2128341  923252049_38877790223  748958041_38511308408_739982   
331891    2128341  923252049_38877790223  748958041_38511308408_739982   
331907    2128341  923252049_38877790223  748958041_38511308408_7399

In [378]:
backward_percentage = (
    backward_visits.sum() / len(visit_sequences) * 100
).round(2)

print("Visits with backward steps:", backward_visits.sum())
print("Total visits:", len(visit_sequences))
print("Percentage of visits with backward steps:", backward_percentage, "%")

Visits with backward steps: 6573
Total visits: 27757
Percentage of visits with backward steps: 23.68 %


In [379]:
def has_repeated_step(steps):
    return len(steps) != len(set(steps))

repeated_visits = visit_sequences.apply(has_repeated_step)

print("Visits with repeated steps:", repeated_visits.sum())
print("Total visits:", len(visit_sequences))
print(
    "Percentage with repeated steps:",
    (repeated_visits.sum() / len(visit_sequences) * 100).round(2),
    "%"
)

Visits with repeated steps: 11346
Total visits: 27757
Percentage with repeated steps: 40.88 %


In [380]:
repeated_step_counts = (
    visit_sequences
    .apply(lambda steps: [step for step in set(steps) if steps.count(step) > 1])
)

print("Repeated steps by visit:")
print(repeated_step_counts.head(10))

Repeated steps by visit:
visit_id
100061345_677868293_127105                    []
100065579_58979991824_496649                  []
100120278_95743019653_642729           [confirm]
100124986_29622874577_527522                  []
100140480_55852036378_404718    [step_3, step_2]
100154767_60541320490_163313     [start, step_1]
100165170_70624318404_468222                  []
100215152_57349471249_380824    [step_2, step_3]
100219420_22495116984_915662             [start]
100250179_91945603763_367112             [start]
Name: process_step, dtype: object


In [381]:
repeated_step_frequency = (
    repeated_step_counts
    .explode()
    .value_counts()
)

print("Most frequently repeated steps:")
print(repeated_step_frequency)

Most frequently repeated steps:
process_step
start      8268
step_1     4605
step_2     3165
step_3     2189
confirm    1376
Name: count, dtype: int64


In [382]:
clean_sequence_visits = visit_sequences[
    ~backward_visits
]

percentage_clean = round(
    len(clean_sequence_visits) / len(visit_sequences) * 100,
    2
)

print("Visits without backward steps:", len(clean_sequence_visits))
print("Percentage without backward steps:", percentage_clean, "%")

Visits without backward steps: 21184
Percentage without backward steps: 76.32 %


In [383]:
completion_by_backward = pd.DataFrame({
    "backward_steps": backward_visits,
    "completed": completed_visits
})

print(
    completion_by_backward
    .groupby("backward_steps")["completed"]
    .agg(["count", "sum", "mean"])
)

                count    sum      mean
backward_steps                        
False           21184  12093  0.570855
True             6573   3096  0.471018


In [384]:
backward_transitions = []

for steps in visit_sequences:
    positions = [step_order[step] for step in steps]
    
    for i in range(1, len(positions)):
        if positions[i] < positions[i - 1]:
            backward_transitions.append(
                (steps[i - 1], steps[i])
            )

backward_transition_counts = pd.Series(
    backward_transitions
).value_counts()

print("Most common backward transitions:")
print(backward_transition_counts)

Most common backward transitions:
(step_1, start)      3539
(step_3, step_2)     1823
(step_2, step_1)     1758
(step_3, start)      1322
(step_2, start)       968
(step_3, step_1)      429
(confirm, start)      329
(confirm, step_1)      69
(confirm, step_3)      59
(confirm, step_2)       2
Name: count, dtype: int64


In [385]:
backward_transition_percentage = (
    backward_transition_counts
    / backward_transition_counts.sum()
    * 100
).round(2)

print("Backward transition distribution (%):")
print(backward_transition_percentage)

Backward transition distribution (%):
(step_1, start)      34.37
(step_3, step_2)     17.70
(step_2, step_1)     17.07
(step_3, start)      12.84
(step_2, start)       9.40
(step_3, step_1)      4.17
(confirm, start)      3.19
(confirm, step_1)     0.67
(confirm, step_3)     0.57
(confirm, step_2)     0.02
Name: count, dtype: float64


In [386]:
ab_summary = (
    experiment_clients
    .groupby("Variation")["client_id"]
    .nunique()
)

print(ab_summary)

Variation
Control    23532
Test       26968
Name: client_id, dtype: int64


In [387]:
client_conversion = (
    web_with_variation
    .groupby(["client_id", "Variation"])["process_step"]
    .apply(lambda steps: "confirm" in steps)
    .reset_index(name="converted")
)

print(client_conversion.head())

Empty DataFrame
Columns: [client_id, Variation, converted]
Index: []


In [388]:
print("client_conversion shape:", client_conversion.shape)
print("\nVariation:")
print(client_conversion["Variation"].value_counts(dropna=False))

print("\nConverted:")
print(client_conversion["converted"].value_counts(dropna=False))

client_conversion shape: (0, 3)

Variation:
Series([], Name: count, dtype: int64)

Converted:
Series([], Name: count, dtype: int64)


In [389]:
print("WEB CLIENT IDs - first 10:")
print(missing_web_data_sorted["client_id"].head(10).tolist())

print("\nEXPERIMENT CLIENT IDs - first 10:")
print(experiment_clients["client_id"].head(10).tolist())

print("\nWEB unique clients:")
print(missing_web_data_sorted["client_id"].nunique())

print("\nEXPERIMENT unique clients:")
print(experiment_clients["client_id"].nunique())

print("\nWEB client ID range:")
print(
    missing_web_data_sorted["client_id"].min(),
    missing_web_data_sorted["client_id"].max()
)

print("\nEXPERIMENT client ID range:")
print(
    experiment_clients["client_id"].min(),
    experiment_clients["client_id"].max()
)

WEB CLIENT IDs - first 10:
[1748647, 2848373, 2848373, 2848373, 2848373, 2848373, 4456733, 4456733, 9156914, 8964191]

EXPERIMENT CLIENT IDs - first 10:
[9988021, 8320017, 4033851, 1982004, 9294070, 9466803, 3170143, 6651403, 5887227, 2105948]

WEB unique clients:
20109

EXPERIMENT unique clients:
70609

WEB client ID range:
169 9999839

EXPERIMENT client ID range:
169 9999839


In [390]:
web_clients = set(missing_web_data_sorted["client_id"].unique())

experiment_clients_with_variation = set(
    experiment_clients.loc[
        experiment_clients["Variation"].notna(),
        "client_id"
    ].unique()
)

matching_clients = web_clients.intersection(
    experiment_clients_with_variation
)

print("Web clients:", len(web_clients))
print("Experiment clients with Variation:", len(experiment_clients_with_variation))
print("Matching clients:", len(matching_clients))

Web clients: 20109
Experiment clients with Variation: 50500
Matching clients: 0


In [392]:
client_conversion = (
    web_with_variation
    .groupby(["client_id", "Variation"])["process_step"]
    .apply(lambda steps: "confirm" in steps.values)
    .reset_index(name="converted")
)

print(client_conversion.shape)
print(client_conversion.head())
print("\nVariation:")
print(client_conversion["Variation"].value_counts(dropna=False))

(0, 3)
Empty DataFrame
Columns: [client_id, Variation, converted]
Index: []

Variation:
Series([], Name: count, dtype: int64)


In [397]:
web_with_variation = web_data.merge(
    experiment_clients[["client_id", "Variation"]],
    on="client_id",
    how="inner"
)

print("web_data shape:", web_data.shape)
print("experiment_clients shape:", experiment_clients.shape)
print("web_with_variation shape:", web_with_variation.shape)

print("\nWeb client_id dtype:")
print(web_data["client_id"].dtype)

print("\nExperiment client_id dtype:")
print(experiment_clients["client_id"].dtype)

print("\nVariation:")
print(web_with_variation["Variation"].value_counts(dropna=False))

web_data shape: (755405, 5)
experiment_clients shape: (70609, 2)
web_with_variation shape: (449831, 6)

Web client_id dtype:
int64

Experiment client_id dtype:
int64

Variation:
Variation
Test       177847
Control    143462
NaN        128522
Name: count, dtype: int64


In [402]:
web_with_variation = web_data.merge(
    experiment_clients.loc[
        experiment_clients["Variation"].notna(),
        ["client_id", "Variation"]
    ],
    on="client_id",
    how="inner"
)

print("web_with_variation shape:", web_with_variation.shape)
print(web_with_variation["Variation"].value_counts())

web_with_variation shape: (321309, 6)
Variation
Test       177847
Control    143462
Name: count, dtype: int64


In [403]:
web_with_variation = web_data.merge(
    experiment_clients.loc[
        experiment_clients["Variation"].notna(),
        ["client_id", "Variation"]
    ],
    on="client_id",
    how="inner"
)

print("web_with_variation shape:", web_with_variation.shape)
print(web_with_variation["Variation"].value_counts())

web_with_variation shape: (321309, 6)
Variation
Test       177847
Control    143462
Name: count, dtype: int64


In [405]:
ab_conversion = (
    client_conversion
    .groupby("Variation")["converted"]
    .agg(
        clients="size",
        conversions="sum"
    )
)

ab_conversion["conversion_rate_%"] = (
    ab_conversion["conversions"] /
    ab_conversion["clients"] * 100
).round(2)

print(ab_conversion)

Empty DataFrame
Columns: [clients, conversions, conversion_rate_%]
Index: []


In [406]:
print("Web clients:", missing_web_data["client_id"].nunique())

print(
    "Web clients that belong to Test/Control:",
    missing_web_data["client_id"].isin(
        experiment_clients.loc[
            experiment_clients["Variation"].notna(),
            "client_id"
        ]
    ).sum()
)

Web clients: 20109
Web clients that belong to Test/Control: 0


In [407]:
print("client_data shape:", client_data.shape)
print("\nclient_data columns:")
print(client_data.columns.tolist())

print("\nUnique clients in client_data:")
print(client_data["client_id"].nunique())

print("\nClient IDs matching Test/Control:")
print(
    client_data["client_id"].isin(
        experiment_clients.loc[
            experiment_clients["Variation"].notna(),
            "client_id"
        ]
    ).sum()
)

client_data shape: (81, 5)

client_data columns:
['client_id', 'visitor_id', 'visit_id', 'process_step', 'date_time']

Unique clients in client_data:
1

Client IDs matching Test/Control:
0


In [408]:
import pandas as pd

experiment_ids = set(
    experiment_clients.loc[
        experiment_clients["Variation"].notna(),
        "client_id"
    ]
)

print("DataFrames containing client_id:")
print("-" * 60)

for name, df in globals().items():
    if isinstance(df, pd.DataFrame) and "client_id" in df.columns:
        matching = df["client_id"].isin(experiment_ids).sum()
        
        print(
            f"{name}: "
            f"shape={df.shape}, "
            f"unique_clients={df['client_id'].nunique()}, "
            f"Test/Control matches={matching}"
        )

DataFrames containing client_id:
------------------------------------------------------------
_: shape=(126662, 5), unique_clients=20109, Test/Control matches=0
___: shape=(5, 5), unique_clients=1, Test/Control matches=0
demo: shape=(70609, 9), unique_clients=70609, Test/Control matches=50500
experiment_clients: shape=(70609, 2), unique_clients=70609, Test/Control matches=50500
web_data_pt1: shape=(343141, 5), unique_clients=58391, Test/Control matches=241228
web_data_pt2: shape=(412264, 5), unique_clients=67430, Test/Control matches=80081
missing_variation: shape=(20109, 2), unique_clients=20109, Test/Control matches=0
web_data_pt1_clean: shape=(341046, 5), unique_clients=58391, Test/Control matches=239985
web_data_pt2_clean: shape=(403595, 5), unique_clients=67430, Test/Control matches=77250
web_data_clean: shape=(744641, 5), unique_clients=120157, Test/Control matches=317235
missing_web_data: shape=(126662, 5), unique_clients=20109, Test/Control matches=0
_15: shape=(5, 9), unique_c

In [409]:
client_conversion = (
    web_with_variation
    .groupby(["client_id", "Variation"])["process_step"]
    .apply(lambda steps: "confirm" in steps.values)
    .reset_index(name="converted")
)

print(client_conversion.head())
print("\nShape:", client_conversion.shape)
print("\nVariation:")
print(client_conversion["Variation"].value_counts())

print("\nConversion:")
print(client_conversion["converted"].value_counts())

   client_id Variation  converted
0        555      Test       True
1        647      Test       True
2        934      Test      False
3       1028   Control      False
4       1104   Control      False

Shape: (50500, 3)

Variation:
Variation
Test       26968
Control    23532
Name: count, dtype: int64

Conversion:
converted
True     34121
False    16379
Name: count, dtype: int64


In [410]:
ab_conversion = (
    client_conversion
    .groupby("Variation")["converted"]
    .agg(
        clients="size",
        conversions="sum"
    )
)

ab_conversion["conversion_rate_%"] = (
    ab_conversion["conversions"]
    / ab_conversion["clients"]
    * 100
).round(2)

print(ab_conversion)

           clients  conversions  conversion_rate_%
Variation                                         
Control      23532        15434              65.59
Test         26968        18687              69.29


In [411]:
from statsmodels.stats.proportion import proportions_ztest

conversions = ab_conversion["conversions"].values
clients = ab_conversion["clients"].values

z_stat, p_value = proportions_ztest(
    conversions,
    clients
)

print("Z-statistic:", round(z_stat, 4))
print("P-value:", round(p_value, 6))

Z-statistic: -8.8745
P-value: 0.0


In [3]:
import pandas as pd

demo = pd.read_csv("../data/raw/df_final_demo.txt", sep=",")
experiment_clients = pd.read_csv(
    "../data/raw/df_final_experiment_clients.txt",
    sep=","
)
web_data_pt1 = pd.read_csv(
    "../data/raw/df_final_web_data_pt_1.txt",
    sep=","
)
web_data_pt2 = pd.read_csv(
    "../data/raw/df_final_web_data_pt_2.txt",
    sep=","
)

print("Demo:", demo.shape)
print("Experiment clients:", experiment_clients.shape)
print("Web pt1:", web_data_pt1.shape)
print("Web pt2:", web_data_pt2.shape)

Demo: (70609, 9)
Experiment clients: (70609, 2)
Web pt1: (343141, 5)
Web pt2: (412264, 5)


In [5]:
web_data = pd.concat(
    [web_data_pt1, web_data_pt2],
    ignore_index=True
)

print("Web data shape:", web_data.shape)
print("Unique clients:", web_data["client_id"].nunique())

Web data shape: (755405, 5)
Unique clients: 120157


In [9]:
web_with_variation = web_data.merge(
    experiment_clients,
    on="client_id",
    how="left"
)

print("web_with_variation shape:", web_with_variation.shape)
print(web_with_variation["Variation"].value_counts(dropna=False))

web_with_variation shape: (755405, 6)
Variation
NaN        434096
Test       177847
Control    143462
Name: count, dtype: int64


In [10]:
print(web_with_variation.columns.tolist())
print(web_with_variation[["client_id", "visit_id", "Variation", "process_step", "date_time"]].head())

['client_id', 'visitor_id', 'visit_id', 'process_step', 'date_time', 'Variation']
   client_id                      visit_id Variation process_step  \
0    9988021  781255054_21935453173_531117      Test       step_3   
1    9988021  781255054_21935453173_531117      Test       step_2   
2    9988021  781255054_21935453173_531117      Test       step_3   
3    9988021  781255054_21935453173_531117      Test       step_2   
4    9988021  781255054_21935453173_531117      Test       step_3   

             date_time  
0  2017-04-17 15:27:07  
1  2017-04-17 15:26:51  
2  2017-04-17 15:19:22  
3  2017-04-17 15:19:13  
4  2017-04-17 15:18:04  


In [11]:
# Make a copy of the experiment data
time_data = web_with_variation[
    web_with_variation["Variation"].isin(["Test", "Control"])
].copy()

# Convert date_time to datetime
time_data["date_time"] = pd.to_datetime(time_data["date_time"])

# Sort each visit chronologically
time_data = time_data.sort_values(
    ["visit_id", "date_time"]
)

# Get the next action and its timestamp within each visit
time_data["next_step"] = (
    time_data.groupby("visit_id")["process_step"].shift(-1)
)

time_data["next_time"] = (
    time_data.groupby("visit_id")["date_time"].shift(-1)
)

# Calculate time spent until the next action
time_data["time_spent_seconds"] = (
    time_data["next_time"] - time_data["date_time"]
).dt.total_seconds()

# Remove the last action of each visit
time_data = time_data.dropna(subset=["time_spent_seconds"])

print("Time data shape:", time_data.shape)
print(time_data.head())

Time data shape: (252104, 9)
        client_id             visitor_id                      visit_id  \
108614    3561384   451664975_1722933822  100012776_37918976071_457913   
240562    7338123  612065484_94198474375   100019538_17884295066_43909   
240561    7338123  612065484_94198474375   100019538_17884295066_43909   
240560    7338123  612065484_94198474375   100019538_17884295066_43909   
240559    7338123  612065484_94198474375   100019538_17884295066_43909   

       process_step           date_time Variation next_step  \
108614      confirm 2017-04-26 13:22:17      Test   confirm   
240562        start 2017-04-09 16:20:56      Test    step_1   
240561       step_1 2017-04-09 16:21:12      Test    step_2   
240560       step_2 2017-04-09 16:21:21      Test    step_1   
240559       step_1 2017-04-09 16:21:35      Test    step_1   

                 next_time  time_spent_seconds  
108614 2017-04-26 13:23:09                52.0  
240562 2017-04-09 16:21:12                16.0  


In [12]:
average_time_by_step = (
    time_data
    .groupby(["Variation", "process_step"])["time_spent_seconds"]
    .mean()
    .reset_index()
)

# Convert seconds to minutes
average_time_by_step["time_spent_minutes"] = (
    average_time_by_step["time_spent_seconds"] / 60
).round(2)

print(average_time_by_step)

  Variation process_step  time_spent_seconds  time_spent_minutes
0   Control      confirm          164.066063                2.73
1   Control        start           62.886243                1.05
2   Control       step_1           50.294342                0.84
3   Control       step_2           91.617708                1.53
4   Control       step_3          135.463130                2.26
5      Test      confirm          243.120249                4.05
6      Test        start           60.538552                1.01
7      Test       step_1           60.580786                1.01
8      Test       step_2           88.612506                1.48
9      Test       step_3          128.924433                2.15


In [13]:
# Sort data chronologically within each visit
error_data = web_with_variation[
    web_with_variation["Variation"].isin(["Test", "Control"])
].copy()

error_data["date_time"] = pd.to_datetime(error_data["date_time"])

error_data = error_data.sort_values(
    ["visit_id", "date_time"]
)

# Define the order of the process steps
step_order = {
    "start": 0,
    "step_1": 1,
    "step_2": 2,
    "step_3": 3,
    "confirm": 4
}

error_data["step_order"] = error_data["process_step"].map(step_order)

# Previous step within each visit
error_data["previous_step_order"] = (
    error_data.groupby("visit_id")["step_order"].shift(1)
)

# A backward transition occurs when the current step
# is earlier than the previous step
error_data["backward_transition"] = (
    error_data["step_order"] < error_data["previous_step_order"]
)

print(
    error_data["backward_transition"].value_counts(dropna=False)
)

backward_transition
False    295220
True      26089
Name: count, dtype: int64


In [14]:
error_rate = (
    error_data
    .groupby("Variation")["backward_transition"]
    .agg(
        total_steps="count",
        errors="sum"
    )
)

error_rate["error_rate_%"] = (
    error_rate["errors"] / error_rate["total_steps"] * 100
).round(2)

print(error_rate)

           total_steps  errors  error_rate_%
Variation                                   
Control         143462    9734          6.79
Test            177847   16355          9.20


In [15]:
from statsmodels.stats.proportion import proportions_ztest

error_counts = error_rate["errors"].values
error_totals = error_rate["total_steps"].values

z_error, p_error = proportions_ztest(
    error_counts,
    error_totals
)

print("Error Rate Z-statistic:", round(z_error, 4))
print("Error Rate P-value:", p_error)

Error Rate Z-statistic: -24.8745
Error Rate P-value: 1.4035488084300064e-136


In [18]:
import pandas as pd

ab_conversion = pd.DataFrame({
    "clients": [23532, 26968],
    "conversions": [15434, 18687]
}, index=["Control", "Test"])

ab_conversion["conversion_rate_%"] = (
    ab_conversion["conversions"]
    / ab_conversion["clients"]
    * 100
).round(2)

print(ab_conversion)

         clients  conversions  conversion_rate_%
Control    23532        15434              65.59
Test       26968        18687              69.29


In [20]:
control_rate = (
    ab_conversion.loc["Control", "conversions"]
    / ab_conversion.loc["Control", "clients"]
)

test_rate = (
    ab_conversion.loc["Test", "conversions"]
    / ab_conversion.loc["Test", "clients"]
)

threshold_rate = control_rate * 1.05

print("Control completion rate:", round(control_rate * 100, 2), "%")
print("Test completion rate:", round(test_rate * 100, 2), "%")
print("5% threshold:", round(threshold_rate * 100, 2), "%")
print(
    "Observed improvement:",
    round((test_rate - control_rate) * 100, 2),
    "percentage points"
)
print(
    "Relative improvement:",
    round((test_rate / control_rate - 1) * 100, 2),
    "%"
)

Control completion rate: 65.59 %
Test completion rate: 69.29 %
5% threshold: 68.87 %
Observed improvement: 3.71 percentage points
Relative improvement: 5.65 %


In [22]:
from statsmodels.stats.proportion import proportions_ztest

# 5% threshold applied to the Control completion rate
threshold_count = control_rate * 1.05

# Test whether the Test group completion rate is greater than the threshold
z_threshold = (
    test_rate - threshold_count
) / (
    (test_rate * (1 - test_rate) / ab_conversion.loc["Test", "clients"]) ** 0.5
)

from scipy.stats import norm

p_threshold = 1 - norm.cdf(z_threshold)

print("Threshold:", round(threshold_count * 100, 2), "%")
print("Test completion rate:", round(test_rate * 100, 2), "%")
print("Z-statistic:", round(z_threshold, 4))
print("P-value:", p_threshold)

Threshold: 68.87 %
Test completion rate: 69.29 %
Z-statistic: 1.5187
P-value: 0.0644202007776894


In [23]:
group_sizes = (
    ab_conversion[["clients"]]
    .copy()
)

group_sizes["percentage"] = (
    group_sizes["clients"]
    / group_sizes["clients"].sum()
    * 100
).round(2)

print(group_sizes)

         clients  percentage
Control    23532        46.6
Test       26968        53.4


In [24]:
# Merge experiment groups with client demographic data
experiment_demo = experiment_clients.merge(
    demo,
    on="client_id",
    how="left"
)

print("Shape:", experiment_demo.shape)

print("\nVariation:")
print(experiment_demo["Variation"].value_counts())

print("\nAge by group:")
print(
    experiment_demo
    .groupby("Variation")["clnt_age"]
    .agg(["count", "mean", "median", "min", "max"])
)

print("\nTenure by group:")
print(
    experiment_demo
    .groupby("Variation")["clnt_tenure_yr"]
    .agg(["count", "mean", "median", "min", "max"])
)

Shape: (70609, 10)

Variation:
Variation
Test       26968
Control    23532
Name: count, dtype: int64

Age by group:
           count       mean  median   min   max
Variation                                      
Control    23526  47.498427    48.5  17.0  96.0
Test       26961  47.163922    47.5  17.0  96.0

Tenure by group:
           count       mean  median  min   max
Variation                                     
Control    23527  12.087686    11.0  2.0  55.0
Test       26961  11.982901    11.0  2.0  55.0


In [25]:
gender_distribution = pd.crosstab(
    experiment_demo["Variation"],
    experiment_demo["gendr"],
    normalize="index"
) * 100

print(gender_distribution.round(2))

gendr          F      M      U     X
Variation                           
Control    32.06  33.88  34.06  0.00
Test       32.33  33.30  34.37  0.01


In [26]:
gender_counts = pd.crosstab(
    experiment_demo["Variation"],
    experiment_demo["gendr"]
)

print(gender_counts)

gendr         F     M     U  X
Variation                     
Control    7543  7970  8014  0
Test       8716  8977  9266  2


In [27]:
from scipy.stats import chi2_contingency

chi2, p_value, dof, expected = chi2_contingency(gender_counts)

print("Chi-square statistic:", round(chi2, 4))
print("P-value:", p_value)

Chi-square statistic: 3.6235
P-value: 0.30509759398220404


In [28]:
from scipy.stats import ttest_ind

control_age = experiment_demo.loc[
    experiment_demo["Variation"] == "Control",
    "clnt_age"
].dropna()

test_age = experiment_demo.loc[
    experiment_demo["Variation"] == "Test",
    "clnt_age"
].dropna()

t_stat, p_value = ttest_ind(
    test_age,
    control_age,
    equal_var=False
)

print("Control mean age:", round(control_age.mean(), 2))
print("Test mean age:", round(test_age.mean(), 2))
print("T-statistic:", round(t_stat, 4))
print("P-value:", p_value)

Control mean age: 47.5
Test mean age: 47.16
T-statistic: -2.4161
P-value: 0.015692719461388675


# Experiment Evaluation

## 1. Was the experiment well-structured?

The experiment was structured as an A/B test with two groups:

- **Control:** 23,532 clients (46.6%)
- **Test:** 26,968 clients (53.4%)

Both groups were exposed to different versions of the digital process:

- Control → old design
- Test → new design

The experiment ran from **March 15, 2017 to June 20, 2017**, providing several weeks of user interaction data.

The experiment structure was appropriate for comparing the performance of the two designs using completion rate, time spent on each step and error rate.

However, the groups were not perfectly equal in size. The Test group represented 53.4% of the sample while the Control group represented 46.6%. Although this imbalance is not necessarily problematic for an A/B test, it should be considered when evaluating the experiment design.

---

## 2. Were clients randomly and equally divided?

The available data shows that the groups were relatively balanced but not equally divided:

| Group | Clients | Percentage |
|---|---:|---:|
| Control | 23,532 | 46.6% |
| Test | 26,968 | 53.4% |

The data does not provide enough information to independently confirm that clients were randomly assigned to the two groups.

Therefore, randomization cannot be fully verified from the available datasets alone.

The demographic analysis also identified a statistically significant difference in mean age between the groups:

- Control mean age: **47.50 years**
- Test mean age: **47.16 years**
- P-value: **0.0157**

Although statistically significant, the difference is only **0.34 years**, suggesting that the practical impact of this difference is very small.

Gender distribution, on the other hand, did not show a statistically significant difference between the groups:

- Chi-square statistic: **3.6235**
- P-value: **0.3051**

Therefore, gender does not appear to represent a significant imbalance between the Test and Control groups.

---

## 3. Were there potential biases?

Some potential sources of bias should be considered.

First, the Test and Control groups were not exactly equal in size.

Second, the age distribution showed a statistically significant difference between the groups. Although the difference was very small in practical terms, age could potentially influence how clients interact with the digital process.

Another limitation is the presence of clients in the web dataset who were not assigned to either experiment group. These observations were excluded from the main A/B test analysis because they did not have a Test or Control variation.

Therefore, the analysis focuses on clients for whom experiment assignment could be identified.

Other potential biases could include differences in client characteristics, previous digital experience, device type, or other factors that are not available in the provided datasets.

---

## 4. Was the experiment duration adequate?

The experiment ran from **March 15, 2017 to June 20, 2017**, covering approximately three months.

This provides a substantial amount of user interaction data and includes a large number of clients and web interactions.

Based on the available data, the duration appears sufficient to generate meaningful insights into completion rates, time spent on process steps and error rates.

However, without information about expected traffic patterns, seasonality, campaign activity or changes in external conditions during the experiment, it is not possible to determine whether the duration completely controlled for all possible time-related effects.

---

## 5. What additional data would improve the analysis?

The analysis could be strengthened by collecting additional information such as:

- Device type and screen size
- Operating system and browser
- Previous experience with Vanguard's digital platform
- Client digital engagement history
- Traffic source or marketing channel
- Geographic location
- More detailed client segmentation
- Technical errors or system failures
- Page loading times
- Customer support interactions during the experiment
- Revenue or business value generated after completion

This additional information could help identify potential confounding factors and provide a better understanding of why users behave differently between the Test and Control groups.

---

## Overall Experiment Evaluation

Overall, the experiment provides a strong basis for evaluating the new digital design because it contains a large sample of clients and detailed records of their interactions with the digital process.

The main limitations are the unequal group sizes, the inability to independently verify random assignment from the available data, and the small but statistically significant difference in average age between the groups.

These limitations should be considered when interpreting the A/B test results, but they do not invalidate the analysis.

In [5]:
import pandas as pd

demo = pd.read_csv(
    "../data/raw/df_final_demo.txt",
    sep=","
)

experiment_clients = pd.read_csv(
    "../data/raw/df_final_experiment_clients.txt",
    sep=","
)

web_data_pt1 = pd.read_csv(
    "../data/raw/df_final_web_data_pt_1.txt",
    sep=","
)

web_data_pt2 = pd.read_csv(
    "../data/raw/df_final_web_data_pt_2.txt",
    sep=","
)

print("Demo:", demo.shape)
print("Experiment clients:", experiment_clients.shape)
print("Web pt1:", web_data_pt1.shape)
print("Web pt2:", web_data_pt2.shape)

Demo: (70609, 9)
Experiment clients: (70609, 2)
Web pt1: (343141, 5)
Web pt2: (412264, 5)


In [8]:
web_data = pd.concat(
    [web_data_pt1, web_data_pt2],
    ignore_index=True
)

print("Web data shape:", web_data.shape)
print("Unique clients:", web_data["client_id"].nunique())

Web data shape: (755405, 5)
Unique clients: 120157


In [9]:
web_with_variation = web_data.merge(
    experiment_clients,
    on="client_id",
    how="left"
)

print("web_with_variation shape:", web_with_variation.shape)
print(web_with_variation["Variation"].value_counts(dropna=False))

web_with_variation shape: (755405, 6)
Variation
NaN        434096
Test       177847
Control    143462
Name: count, dtype: int64


In [12]:
client_conversion = (
    web_with_variation
    .groupby(["client_id", "Variation"])["process_step"]
    .apply(lambda steps: "confirm" in steps.values)
    .reset_index(name="converted")
)

print(client_conversion.head())
print(client_conversion.shape)

print("\nVariation:")
print(client_conversion["Variation"].value_counts())

print("\nConversion:")
print(client_conversion["converted"].value_counts())

   client_id Variation  converted
0        555      Test       True
1        647      Test       True
2        934      Test      False
3       1028   Control      False
4       1104   Control      False
(50500, 3)

Variation:
Variation
Test       26968
Control    23532
Name: count, dtype: int64

Conversion:
converted
True     34121
False    16379
Name: count, dtype: int64


In [13]:
ab_conversion = (
    client_conversion
    .groupby("Variation")["converted"]
    .agg(
        clients="size",
        conversions="sum"
    )
)

ab_conversion["conversion_rate_%"] = (
    ab_conversion["conversions"]
    / ab_conversion["clients"]
    * 100
).round(2)

print(ab_conversion)

           clients  conversions  conversion_rate_%
Variation                                         
Control      23532        15434              65.59
Test         26968        18687              69.29


In [14]:
tableau_kpis = ab_conversion.reset_index()

tableau_kpis.to_csv(
    "../data/processed/tableau_kpis.csv",
    index=False
)

print(tableau_kpis)

  Variation  clients  conversions  conversion_rate_%
0   Control    23532        15434              65.59
1      Test    26968        18687              69.29


In [16]:
tableau_kpis = ab_conversion.reset_index()

tableau_kpis

,Variation,clients,conversions,conversion_rate_%
0,Control,23532,15434,65.59
1,Test,26968,18687,69.29


In [20]:
time_data = web_with_variation.copy()

time_data["date_time"] = pd.to_datetime(time_data["date_time"])

time_data = time_data.sort_values(
    ["visit_id", "date_time"]
)

time_data["next_step"] = (
    time_data.groupby("visit_id")["process_step"].shift(-1)
)

time_data["next_time"] = (
    time_data.groupby("visit_id")["date_time"].shift(-1)
)

time_data["time_spent_seconds"] = (
    time_data["next_time"] - time_data["date_time"]
).dt.total_seconds()

# Remove the last action of each visit
time_data = time_data.dropna(subset=["time_spent_seconds"])

print("Time data shape:", time_data.shape)
print(time_data.head())

Time data shape: (597310, 9)
        client_id             visitor_id                      visit_id  \
108614    3561384   451664975_1722933822  100012776_37918976071_457913   
447951    9056452  306992881_89423906595     1000165_4190026492_760066   
447950    9056452  306992881_89423906595     1000165_4190026492_760066   
447949    9056452  306992881_89423906595     1000165_4190026492_760066   
447948    9056452  306992881_89423906595     1000165_4190026492_760066   

       process_step           date_time Variation next_step  \
108614      confirm 2017-04-26 13:22:17      Test   confirm   
447951        start 2017-06-04 01:07:29       NaN    step_1   
447950       step_1 2017-06-04 01:07:32       NaN    step_2   
447949       step_2 2017-06-04 01:07:56       NaN    step_3   
447948       step_3 2017-06-04 01:09:13       NaN   confirm   

                 next_time  time_spent_seconds  
108614 2017-04-26 13:23:09                52.0  
447951 2017-06-04 01:07:32                 3.0  


In [21]:
average_time_by_step = (
    time_data
    .groupby(["Variation", "process_step"])["time_spent_seconds"]
    .mean()
    .reset_index()
)

average_time_by_step["time_spent_minutes"] = (
    average_time_by_step["time_spent_seconds"] / 60
).round(2)

print(average_time_by_step)

  Variation process_step  time_spent_seconds  time_spent_minutes
0   Control      confirm          171.934361                2.87
1   Control        start           62.941832                1.05
2   Control       step_1           50.269672                0.84
3   Control       step_2           91.617708                1.53
4   Control       step_3          135.567522                2.26
5      Test      confirm          248.060530                4.13
6      Test        start           60.561098                1.01
7      Test       step_1           60.609474                1.01
8      Test       step_2           88.612506                1.48
9      Test       step_3          128.916522                2.15


In [22]:
tableau_time = average_time_by_step.copy()

tableau_time

,Variation,process_step,time_spent_seconds,time_spent_minutes
0,Control,confirm,171.934361,2.87
1,Control,start,62.941832,1.05
2,Control,step_1,50.269672,0.84
3,Control,step_2,91.617708,1.53
4,Control,step_3,135.567522,2.26
5,Test,confirm,248.060530,4.13
6,Test,start,60.561098,1.01
7,Test,step_1,60.609474,1.01
8,Test,step_2,88.612506,1.48
9,Test,step_3,128.916522,2.15


In [23]:
tableau_time = average_time_by_step.copy()

tableau_time

,Variation,process_step,time_spent_seconds,time_spent_minutes
0,Control,confirm,171.934361,2.87
1,Control,start,62.941832,1.05
2,Control,step_1,50.269672,0.84
3,Control,step_2,91.617708,1.53
4,Control,step_3,135.567522,2.26
5,Test,confirm,248.060530,4.13
6,Test,start,60.561098,1.01
7,Test,step_1,60.609474,1.01
8,Test,step_2,88.612506,1.48
9,Test,step_3,128.916522,2.15


In [26]:
error_data = web_with_variation.copy()

# Garantir que as visitas estão ordenadas cronologicamente
error_data["date_time"] = pd.to_datetime(error_data["date_time"])

error_data = error_data.sort_values(
    ["visit_id", "date_time"]
)

# Próximo passo dentro de cada visita
error_data["next_step"] = (
    error_data.groupby("visit_id")["process_step"].shift(-1)
)

# Definir a ordem correta do processo
step_order = {
    "start": 0,
    "step_1": 1,
    "step_2": 2,
    "step_3": 3,
    "confirm": 4
}

error_data["current_order"] = error_data["process_step"].map(step_order)
error_data["next_order"] = error_data["next_step"].map(step_order)

# Identificar movimentos para trás
error_data["backward_transition"] = (
    error_data["next_order"] < error_data["current_order"]
)

print(error_data["backward_transition"].value_counts(dropna=False))

backward_transition
False    691527
True      63878
Name: count, dtype: int64


In [29]:
error_rate = (
    error_data
    .groupby("Variation")["backward_transition"]
    .agg(
        total_steps="count",
        errors="sum"
    )
)

error_rate["error_rate_%"] = (
    error_rate["errors"] / error_rate["total_steps"] * 100
).round(2)

print(error_rate)

           total_steps  errors  error_rate_%
Variation                                   
Control         143462    9782          6.82
Test            177847   16445          9.25


In [35]:
tableau_error = error_rate.reset_index()

tableau_error

,Variation,total_steps,errors,error_rate_%
0,Control,143462,9782,6.82
1,Test,177847,16445,9.25


In [43]:
import os

os.makedirs("../data/processed", exist_ok=True)

tableau_kpis.to_csv(
    "../data/processed/tableau_kpis.csv",
    sep=";",
    index=False
)

tableau_time.to_csv(
    "../data/processed/tableau_time.csv",
    sep=";",
    index=False
)

tableau_error.to_csv(
    "../data/processed/tableau_error.csv",
    sep=";",
    index=False
)

tableau_time.to_csv(
    "../data/processed/tableau_time_by_step.csv",
    sep=";",
    index=False
)

print("CSV files created successfully!")
print(os.listdir("../data/processed"))

CSV files created successfully!
['tableau_error.csv', 'tableau_kpis.csv', 'tableau_time.csv', 'tableau_time_by_step.csv']
